crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -379,6 +393,67 @@ impl RecursorDnsHandle {
             }
         }
 
+        // If we still have no NS records, try to query the parent zone for child NS servers
+        if config_group.is_empty() && !need_ips_for_names.is_empty() {
+            debug!("Priming zone {zone} via parent zone {}", zone.base_name());
+
+            let a_primes = need_ips_for_names
+                .iter()
+                .filter(|name| crate::is_subzone(zone.clone(), name.0.clone()))
+                .take(1)
+                .map(|name| {
+                    nameserver_pool
+                        .lookup(
+                            Query::query(name.0.clone(), RecordType::A),
+                            self.security_aware,
+                        )
+                        .boxed()
+                });
+
+            let aaaa_primes = need_ips_for_names
+                .iter()
+                .filter(|name| crate::is_subzone(zone.clone(), name.0.clone()))
+                .take(1)
+                .map(|name| {
+                    nameserver_pool
+                        .lookup(
+                            Query::query(name.0.clone(), RecordType::AAAA),
+                            self.security_aware,
+                        )
+                        .boxed()
+                });
+
+            let mut primes: Vec<_> = a_primes.chain(aaaa_primes).collect();
+
+            while !primes.is_empty() {
```

Comment:
nit: there's a pretty similar loop in line 356 that would be nice to refactor to avoid repetition

Reply:
That previous loop resolves the futures for the queries for non-child nameservers.  I think the situation where a parent NS returns an in-domain/child NS delegation without glue should be treated as an exception, and we should only attempt to resolve those name servers if there are no other usable records present.  That does require a bit of repetition, though.

Reply:
@marcus0x62 , not sure if you both are talking past each other. But I think the logic here is identical, but the the debug log lines are different. If you think the added debug info is important, it should be fine to pass a const string into a function that performs this conversion and still adds the additional context to the debug info.

Reply:
I'm not really worried about preserving the debug lines, but rather not firing off additional NS queries that may not be necessary and that are likely to end in failure (and prolonging failure in the process.)  I think the case where a parent NS has glue but doesn't return it is pretty unusual, although [RFC 1034](https://datatracker.ietf.org/doc/rfc1034/) leaves the door open to that as an implementation choice.  [RFC 9471](https://datatracker.ietf.org/doc/rfc9471/) (Section 3.1) says resolvers

> "MUST include all available glue records for in-domain name servers in the additional section or MUST set TC=1 if constrained by message size."

Reply:
are we talking about the same thing? these loops look identical:

```rust
             while !a_resolves.is_empty() {
                let (next, _, rest) = select_all(a_resolves).await;
                a_resolves = rest;
                match next {
                    Ok(response) => {
                        debug!("A or AAAA response: {:?}", response);
                        let ips = response.iter().filter_map(RData::ip_addr);
                        for ip in ips {
                            let udp =
                                NameServerConfig::new(SocketAddr::from((ip, 53)), Protocol::Udp);
                            let tcp =
                                NameServerConfig::new(SocketAddr::from((ip, 53)), Protocol::Tcp);
                            config_group.push(udp);
                            config_group.push(tcp);
                        }
                    }
                    Err(e) => {
                        warn!("resolve failed {}", e);
                    }
                }
            }
```

And while we're discussing this, I think we're setting the `NameServerConfig` a little differently here on line 326:

```rust
udp.trust_negative_responses = true;
tcp.trust_negative_responses = true;
```

Do we need that in all of these NameServerConfigs?

Reply:
I think maybe we are.  Here's how I see the old code:

There are two distinct, sequential code blocks used to add entries to the NS pool:

1) An unconditionally fired loop on returned glue records [here](https://github.com/hickory-dns/hickory-dns/blob/e8850b788b5d10279998feb4b74ce3ffbcd1b9d6/crates/recursor/src/recursor_dns_handle.rs#L286)
2) A conditionally fired block that iterates over any NS servers that did not have glue returned [here](https://github.com/hickory-dns/hickory-dns/blob/e8850b788b5d10279998feb4b74ce3ffbcd1b9d6/crates/recursor/src/recursor_dns_handle.rs#L343).  It only runs if a) there are unresolved nameserver addresses from the query to parent NS and if the glue list is empty.

The problem with the second loop is that we attempt to resolve a child or "in-domain" nameserver address that was returned without glue, we'll get stuck in an infinite loop.

The new code has three distinct, sequential blocks:

1) An unconditionally fired loop on returned glue records [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L287). This loop has not changed.
2) A conditionally fired block on any non-child nameservers [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L348) In addition to the original checks (that the config group is empty and that there are records in the need_ips_for_names vec,) this block filters out child nameservers from the delegation set and does not attempt to resolve them.
3. A conditionally fired block on any child nameservers [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L397) that does the same thing as block 2 on child-nameservers, but only if there still aren't any entries in the config group and which doesn't recursively call resolve (to avoid the infinite recursion problem.)

So, I guess ultimately, I don't have a problem moving the inner loops [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L370) and [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L428) used to resolve the lookup futures into a shared function or something, but I don't think it would be a good idea to fold steps two and three *overall* into a single block.

Reply:
> So, I guess ultimately, I don't have a problem moving the inner loops [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L370) and [here](https://github.com/marcus0x62/hickory-dns/blob/845f3e4666e0246ab553b7ee62eb2064c8c3902c/crates/recursor/src/recursor_dns_handle.rs#L428) used to resolve the lookup futures into a shared function or something, 

yes, that's what I was referring to, not sure if @japaric was thinking of something beyond that.

> but I don't think it would be a good idea to fold steps two and three overall into a single block.

agreed.

> I can't think of a reason why we would want to only trust negative responses from nameservers that we learned the addresses of via glue records, which is what the current behavior effectively is.

Agreed. I think we either don't "trust" any, or we use the same setup for each.

Reply:
@bluejekyll I moved what I could from the inner loop into a shared function in [crates/recursor/src/lib.rs](https://github.com/hickory-dns/hickory-dns/blob/ba2862bd88ca9259fc65ed1736852918929409fe/crates/recursor/src/lib.rs#L140), but that ended up just being the innermost iteration over the ip range.  The rest of the code had enough type differences (caused by the differing return types in resolve vs lookup) to prevent an easy collapse into a single function (at least that I could figure out.)

Doing it this way did allow me to fold the config group calls from the glue IP section in as well.

Reply:
@japaric, do you have any open concerns? otherwise, I think we can close out this conversation.